### PR TITLE
Basket feature

### PR DESCRIPTION
**Pull request by VeniNovakov**

### CODE REVIEW

File: pizzeria-frontend/src/Menu.tsx [hunk 1]
```diff
@@ -1,100 +1,136 @@
...
 const Menu = () => {
+  const [selectedFilter, setSelectedFilter] = useState("");
+
+  const handleFilterClick = (filterType: string) => {
+    setSelectedFilter(filterType);
+  };
+
+  const filteredMenu = selectedFilter
+    ? menu.filter((product) => product.category === selectedFilter)
+    : menu;
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/Menu.tsx]: Move this to useEffect with dependancy prop selectedFilter, as this way it would update only on re-render. ( Could still work but unclear )

File: pizzeria-frontend/src/Menu.tsx [hunk 2]
```diff
@@ -1,100 +1,136 @@
...
+        {filteredMenu.map((product) => (
           <Product product={product} key={product.id} />
         ))}
       </div>
     </div>
   );
 };
 
-const Filter = () => {
+const Filter = ({ onFilterClick }: any) => {
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/Menu.tsx]: Remove **any**, you can create an interface with {
 onFilterClick: (_: string) => void;
 }
 and so it knows its getting a function with props and output

File: pizzeria-frontend/src/Menu.tsx [hunk 3]
```diff
@@ -1,100 +1,136 @@
...
+          key={index}
+          onFilterClick={() => onFilterClick(item.name)}
+        />
+      ))}
     </div>
   );
 };
 
-const FilterItem = (props: any) => {
+const FilterItem = ({ item, onFilterClick }: any) => {
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/Menu.tsx]: Same here, remove any

File: pizzeria-frontend/src/Menu.tsx [hunk 4]
```diff
@@ -1,100 +1,136 @@
...
+    );
+  }, [basketItems, setBasketCounter]);
 
   const updateBasket = (product: ProductType) => {
     setBasketCounter(basketCounter + 1);
 
     const newBasketObj: BasketItem = { product: product, amount: 1 };
 
     if (basketItems.length === 0) {
+      console.log(basketItems);
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/Menu.tsx]: might want to remove this, if its intentional just resolve this thread

File: pizzeria-frontend/src/pages/ProductPage.tsx [hunk 5]
```diff
@@ -1,26 +1,208 @@
...
+      }else{
+        Cookies.set('tempProduct', JSON.stringify({...ck, addOns: [...updatedAddOns]}));
+      }
+    }
+    }, [amount])
+
+  const removeAmount = () => {
+    if (amount > 1 && instanceOfProduct(item)) {
+      amount--;
+      setAmount(amount);
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/ProductPage.tsx]: Why are you doing 
`amount--;`
and then
`setAmount(amount); `
Cant you just do `setAmount(amount - 1)` ?

File: pizzeria-frontend/src/pages/ProductPage.tsx [hunk 6]
```diff
@@ -1,26 +1,208 @@
...
+    if (amount > 1 && instanceOfProduct(item)) {
+      amount--;
+      setAmount(amount); 
+    }else if(amount > 0 && !instanceOfProduct(item)){
+      amount--;
+      setAmount(amount); 
+    }
+
+    if(amount === 1 && instanceOfProduct(item)){
+      console.log('dsadsad');
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/ProductPage.tsx]: Add a TODO if this is not part of this PR's task

File: pizzeria-frontend/src/pages/basketPage.tsx [hunk 7]
```diff
@@ -1,35 +1,207 @@
...
+  useEffect(() => {
+    console.log("basketItems:");
+
+    setBasketCounter(basketItems.length);
+
+    const basketString = JSON.stringify(basketItems);
+
+    Cookies.set("basket", basketString);
+
+  }, [basketItems, setBasketCounter]);
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/basketPage.tsx]: Have you tested whether this calls itself permanently as i have a suspicion it might?

File: pizzeria-frontend/src/pages/basketPage.tsx [hunk 8]
```diff
@@ -1,35 +1,207 @@
...
     </div>
   );
 };
-const ProductOrdered = (props: ProductProps) => {
-  return <div className=""></div>;
+
+const ProductOrdered = (props: BasketProps) => {
+  const { basketItems, setBasketItems } = useBasketContent();
+  const { setBasketCounter } = useBasket();
+  const IsButtonDisabled = props.item.amount === 1 ? true:false
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/basketPage.tsx]: Would put it in a useEffect again,
useEffect(() => {
  setIsButtonDisabled(props.item.amount === 1 );
  }, [props.item.amount]);

File: pizzeria-frontend/src/pages/basketPage.tsx [hunk 9]
```diff
@@ -1,35 +1,207 @@
...
+          <button
+            type="button"
+            className="shadow-md hover:shadow-inner hover:bg-slate-100 rounded-md border-slate-400 border"
+          >
+            ORDER
+          </button>
+        </>
+      ) : (
+        <div></div>
+      )}
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/basketPage.tsx]: You could change the ? to &&
````
{basketItems.length && (
        <>
          <Total></Total>
          <button
            type="button"
            className="shadow-md hover:shadow-inner hover:bg-slate-100 rounded-md border-slate-400 border"
          >
            ORDER
          </button>
        </>
      )}

File: pizzeria-frontend/src/pages/ProductPage.tsx [hunk 10]
```diff
@@ -1,26 +1,208 @@
...
+  const { id } = useParams<{ id: string }>();
+  const numericId = parseInt(id as string, 10);
+  const [ productAmount, SetProductAmount ] = useState(1);
+  const {basketItems, setBasketItems} = useBasketContent();
+  const selectedItem: ProductType | undefined = menu.find((item) => item.id === numericId);
+  const submitOrder = () => {
+    //when clicked it will add it to the basket cookie
+    const tempProduct = JSON.parse(Cookies.get('tempProduct') || '');
+    Cookies.set('basket', JSON.stringify([...basketItems, tempProduct]));
+    setBasketItems(JSON.parse(Cookies.get('basket') || '[]'));
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/ProductPage.tsx]: Optional,
As in every place you do setBasketItems, then you do Cookies.set('basket',
It would be a lot of refactoring so thats why im saying its optional, but it would make it clearer code. 
You could add in the Basket Context an use Effect or directly in the setBascetItems to add it to the cookies so you dont have to worry every time to add it

File: pizzeria-frontend/src/pages/ProductPage.tsx [hunk 11]
```diff
@@ -1,26 +1,208 @@
...
+        Cookies.set('tempProduct', JSON.stringify({...ck, addOns: [...updatedAddOns.filter(addOn => addOn !== null)]}));
+        return;
+      }
+      if(JSON.stringify(updatedAddOns) === JSON.stringify(ck.addOns)){
+        Cookies.set('tempProduct', JSON.stringify({...ck, addOns: [...ck.addOns, {...item, amount}]}))
+      }else{
+        Cookies.set('tempProduct', JSON.stringify({...ck, addOns: [...updatedAddOns]}));
+      }
+    }
+    }, [amount])
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/pages/ProductPage.tsx]: This is quite a big useEffect and hard to read, would split it in couple if functions and every if would call its function then return. That way here you have all the logic like
if(Cookies.get('tempProduct')){
  doCookieUpdatedAddOns()
  return;
 }
 And if someone needs to see what the this if does he just goes in the function and looks

File: pizzeria-frontend/src/Menu.tsx (hunk 1 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/Menu.tsx]: done

File: pizzeria-frontend/src/pages/basketPage.tsx (hunk 7 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/pages/basketPage.tsx]: Tested it now, it does not console log endlessly if that means something. I think it could have happened if I changed the basketItems array in this useEffect though

File: pizzeria-frontend/src/pages/ProductPage.tsx (hunk 10 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/pages/ProductPage.tsx]: Done. Chose to add the useEffect in the Basket Context as it sounds like the better option

File: pizzeria-frontend/src/pages/ProductPage.tsx (hunk 11 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/pages/ProductPage.tsx]: Done. In the new commit the useEffect is way smaller and contains much smaller part of the logic. Improved readability(could probably be better also). The logic of updating the amount of product/addOn is now handled by other functions

File: pizzeria-frontend/src/pages/basketPage.tsx (hunk 9 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/pages/basketPage.tsx]: Done

File: pizzeria-frontend/src/pages/basketPage.tsx (hunk 8 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/pages/basketPage.tsx]: Done

File: pizzeria-frontend/src/Menu.tsx (hunk 4 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/Menu.tsx]: Done

File: pizzeria-frontend/src/Menu.tsx (hunk 3 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/Menu.tsx]: Done

File: pizzeria-frontend/src/Menu.tsx (hunk 2 above)
Comment (VeniNovakov) [reply pizzeria-frontend/src/Menu.tsx]: Done

File: pizzeria-frontend/src/pages/basketPage.tsx (hunk 7 above)
Comment (NikolaPavlovski002) [reply pizzeria-frontend/src/pages/basketPage.tsx]: Yup, if it doesnt consoleLog endlessly its good

File: pizzeria-frontend/src/Menu.tsx [hunk 12]
```diff
@@ -1,56 +1,102 @@
...
   );
 };
 
 const Product = (props: ProductProps) => {
-  const { basketCounter, setBasketCounter } = useBasket();
   const { basketItems, setBasketItems } = useBasketContent();
+  const { basketCounter, setBasketCounter } = useBasket();
 
   useEffect(() => {
     Cookies.set("basket", JSON.stringify(basketItems));
```
Comment (NikolaPavlovski002) [inline pizzeria-frontend/src/Menu.tsx:94]: Now that you have the useEffect, you can skip this here

**Comment by NikolaPavlovski002** [comment]
Looks good, you can merge
